### PR TITLE
feat(YAML config): Add expanded when editing an instance

### LINKED ISSUE #1456: [feature]  add `--expanded` option to yaml config view
currently, the view of raw yaml configuration on UI **does not show expanded configuration.**
i understand this may be due to the edition possibility restriction (does not make sense to edit inherited attributes there), but i'm sure we can figure something out from the UX point of view.

i spotted this need on instances but maybe it applies also to other resources.

Image

### PR DESCRIPTION
**Pull request by kimanhou**
## Done

- When editing an instance, add expanded YAML configuration

Fixes #1456

## QA

1. Run the LXD-UI:
    - On the demo server via the link posted by @webteam-app below. This is only available for PRs created by collaborators of the repo. Ask @Kxiru or @edlerd for access.
    - With a local copy of this branch, [build and run as described in the docs](https://github.com/canonical/lxd-ui/blob/main/CONTRIBUTING.md#setting-up-for-development).
2. Perform the following QA steps:
    - Use an instance that inherits from a profile, preferably a profile with a device.
    - Go to the instance configuration, switch to YAML configuration: 2 radio buttons should appear (`Instance only` and `Expanded`)
    - Check `Expanded`. Make sure the YAML is read only with a hover text and `Cancel` and `Save` buttons are not displayed. Make sure you can see the inherited profile configuration.

## Screenshots

image

Medium screen with line break
image

Annotated expanded YAML
image

### CODE REVIEW

File: src/components/forms/YamlExpandedSelector.tsx [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  setIsExpanded: (expanded: boolean) => void;
+}
+
+const YamlExpandedSelector: FC<Props> = ({ isExpanded, setIsExpanded }) => {
+  const isMediumScreen = useIsScreenBelow(mediumScreenBreakpoint);
+
+  return (
+    <div className="yaml-expanded-radio-wrapper">
+      <RadioInput
+        label={isMediumScreen ? "Local" : "Instance only"}
```
Comment (kimanhou) [inline src/components/forms/YamlExpandedSelector.tsx]: Any suggestion for a better text here ?

Comment (vosdev) [reply src/components/forms/YamlExpandedSelector.tsx]: what about:

[ ] Local  [x] Inherited

Comment (kimanhou) [reply src/components/forms/YamlExpandedSelector.tsx]: That could work too actually. I put Expanded to be aligned with the CLI `lxc config show INSTANCE --expanded`

Comment (edlerd) [reply src/components/forms/YamlExpandedSelector.tsx]: Maybe we can have both and use "Expand inherited values"?

Comment (kimanhou) [reply src/components/forms/YamlExpandedSelector.tsx]: Like this ? "Expand inherited values" is too long
image

File: src/pages/instances/EditInstance.tsx [hunk 2]
```diff
@@ -258,49 +269,61 @@ const EditInstance: FC<Props> = ({ instance }) => {
...
+        <footer className="edit-instance-footer bottom-btns">
+          <YamlSwitch
+            formik={formik}
+            section={section}
+            setSection={updateSection}
+            extraControls={
+              <YamlExpandedSelector
+                isExpanded={isYamlExpanded}
+                setIsExpanded={setIsYamlExpanded}
+              />
```
Comment (edlerd) [inline src/pages/instances/EditInstance.tsx]: Can we make this a sibling to the `YamlSwitch`, so we don't need any changes to the other component?

It seems not necessary to push it into the `YamlSwitch`.

File: src/pages/instances/EditInstance.tsx [hunk 3]
```diff
@@ -83,6 +83,7 @@ const EditInstance: FC<Props> = ({ instance }) => {
   const navigate = useNavigate();
   const [version, setVersion] = useState(0);
   const { canEditInstance } = useInstanceEntitlements();
+  const [isYamlExpanded, setIsYamlExpanded] = useState(false);
```
Comment (edlerd) [inline src/pages/instances/EditInstance.tsx]: I confused this initially with "is the yaml editor open". Maybe a slightly different naming can help here. With second option "expandInheritedValues":
```suggestion
  const [yamlType, yamlType] = useState("local");
```

File: src/pages/instances/EditInstance.tsx [hunk 4]
```diff
@@ -158,6 +159,10 @@ const EditInstance: FC<Props> = ({ instance }) => {
       formik.setFieldValue("yaml", undefined);
     }
 
+    if (newSection !== YAML_CONFIGURATION) {
+      setIsYamlExpanded(false);
+    }
+
```
Comment (edlerd) [inline src/pages/instances/EditInstance.tsx]:
```suggestion
```
I think we can keep the current state here as is. A user might switch back and forth, and forcing them always into the local view might not be intended.

File: src/pages/instances/EditInstance.tsx [hunk 5]
```diff
@@ -258,49 +269,61 @@ const EditInstance: FC<Props> = ({ instance }) => {
...
                 setYaml={(yaml) => {
                   ensureEditMode(formik);
                   formik.setFieldValue("yaml", yaml);
                 }}
-                readOnly={!!formik.values.editRestriction}
-                readOnlyMessage={formik.values.editRestriction}
+                readOnly={isYamlExpanded || !!formik.values.editRestriction}
+                readOnlyMessage={
+                  isYamlExpanded
+                    ? "Expanded configuration is read-only"
```
Comment (edlerd) [inline src/pages/instances/EditInstance.tsx:287]: I wanted to suggest this and am happy to see it already in place. 👍

File: src/pages/instances/EditInstance.tsx [hunk 6]
```diff
@@ -258,49 +269,61 @@ const EditInstance: FC<Props> = ({ instance }) => {
...
-              formik={formik}
-              baseUrl={baseUrl}
-              isYaml={section === slugify(YAML_CONFIGURATION)}
-              disabled={hasDiskError(formik) || hasNetworkError(formik)}
-            />
-          </>
-        )}
-      </FormFooterLayout>
+      <div className="p-bottom-controls form-footer" id="form-footer">
+        <footer className="edit-instance-footer bottom-btns">
```
Comment (edlerd) [inline src/pages/instances/EditInstance.tsx]: I wonder if we should keep the `FormFooterLayout` here and override styling as you already do by targeting `.edit-instance .bottom-btns` and its descendants. With the current approach, it becomes a rather bespoke component. Wdyt?

File: src/components/forms/YamlExpandedSelector.tsx [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  mediumScreenBreakpoint,
+  useIsScreenBelow,
+} from "context/useIsScreenBelow";
+
+interface Props {
+  isExpanded: boolean;
+  setIsExpanded: (expanded: boolean) => void;
+}
+
+const YamlExpandedSelector: FC<Props> = ({ isExpanded, setIsExpanded }) => {
```
Comment (edlerd) [inline src/components/forms/YamlExpandedSelector.tsx]:
```suggestion
const YamlTypeSelector: FC<Props> = ({ isExpanded, setIsExpanded }) => {
```
Pending the other naming suggestion -- maybe clearer?

File: src/util/yaml.tsx [hunk 8]
```diff
@@ -7,3 +9,90 @@ export const yamlToObject = (yamlString: string): object => {
...
+        device &&
+        (device as unknown as Record<string, unknown>)[key] !== undefined
+      );
+    }
+    return p?.config?.[key] !== undefined;
+  });
+  return source || "profile";
+};
+
+export const getAnnotatedYaml = (
```
Comment (edlerd) [inline src/util/yaml.tsx]:
```suggestion
export const expandInheritedValuesYaml = (
```
Shall we align this with the visible option?

Comment (edlerd) [inline src/util/yaml.tsx]: Sometimes the comments are on a new line, see [this instance](https://lxd-ui-1828.demos.haus/ui/project/default/instance/DemoInstance/configuration/yaml-configuration). I think the comment should always be on the same line.

Image

File: src/util/yaml.tsx [hunk 9]
```diff
@@ -7,3 +9,90 @@ export const yamlToObject = (yamlString: string): object => {
...
+      // Fully inherited device block
+      const source = reversedProfiles.find(
+        (p) => profiles.find((all) => all.name === p)?.devices?.[name],
+      );
+      sections.push(
+        `  # --- inherited from profile: ${source || "profile"} ---`,
+      );
+      sections.push(`  ${name}:`);
+      sections.push(objectToYaml(config).replace(/^/gm, "    ").trimEnd());
+      sections.push(`  # --- end of inherited ---`);
```
Comment (edlerd) [inline src/util/yaml.tsx]: I think instead of beginning and end comments, better comment directly on the start of the device:
```
  disk-device-1: # inherited from profile: disk
    path: /foo
    pool: pool1
    source: vol1
    type: disk
```

File: src/util/yaml.tsx [hunk 10]
```diff
@@ -7,3 +9,90 @@ export const yamlToObject = (yamlString: string): object => {
...
+    "snapshots",
+    "state",
+    "etag",
+  ];
+  const metadata = Object.fromEntries(
+    Object.entries(instance).filter(([key]) => !exclude.includes(key)),
+  );
+  sections.push(objectToYaml(metadata).trim());
+
+  sections.push("config:");
```
Comment (edlerd) [inline src/util/yaml.tsx]: This generates two config sections where there should be only one:

https://github.com/user-attachments/assets/951003a0-5725-4b6c-871f-a7695d052155

Same for the devices if local and inherited devices are present.

File: src/components/forms/YamlExpandedSelector.tsx (hunk 1 above)
Comment (edlerd) [reply src/components/forms/YamlExpandedSelector.tsx]: It looks okay in the screenshot. No?

File: src/util/yaml.tsx (hunk 8 above)
Comment (kimanhou) [reply src/util/yaml.tsx]: Is it because the network config in cloud-init profile is an empty line ? What to do in that case ?

image

Comment (kimanhou) [reply src/util/yaml.tsx]: I did something
image

File: src/pages/instances/EditInstance.tsx [hunk 11]
```diff
@@ -256,31 +271,41 @@ const EditInstance: FC<Props> = ({ instance }) => {
...
+                }
               >
                 <YamlNotification entity="instance" docPath="/instances" />
               </YamlForm>
             )}
           </Col>
         </Row>
       </Form>
-      <FormFooterLayout>
+      <FormFooterLayout isAlignRight={false}>
```
Comment (edlerd) [inline src/pages/instances/EditInstance.tsx:297]:
```suggestion
      <FormFooterLayout isAlignRight>
```

File: src/pages/instances/EditInstance.tsx [hunk 12]
```diff
@@ -256,31 +271,41 @@ const EditInstance: FC<Props> = ({ instance }) => {
...
           formik={formik}
           section={section}
           setSection={updateSection}
         />
-        {readOnly ? null : (
-          <>
+        {isYamlSection && (
+          <YamlTypeSelector yamlType={yamlType} setYamlType={setYamlType} />
+        )}
+        {readOnly || yamlType === "expandInheritedValues" ? null : (
```
Comment (edlerd) [inline src/pages/instances/EditInstance.tsx]:
```suggestion
        {readOnly ? null : (
```
When we go to the `expandInheritedValues`, we should also cancel the current YAML editing. That simplifies the case here, but pending changes will be discarded. That pending changes will be discarded is not necessarily expected. So I think we should also show the warning if any pending changes are present, like when going into YAML edit, starting to change, and leaving YAML mode.

File: src/components/forms/YamlTypeSelector.tsx [hunk 13]
```diff
@@ -0,0 +1,30 @@
+import type { FC } from "react";
+import { RadioInput } from "@canonical/react-components";
+
+interface Props {
+  yamlType: "local" | "expandInheritedValues";
+  setYamlType: (yamlType: "local" | "expandInheritedValues") => void;
```
Comment (edlerd) [inline src/components/forms/YamlTypeSelector.tsx]:
```suggestion
export type YamlType = "local" | "expandInheritedValues";

interface Props {
  yamlType: YamlType;
  setYamlType: (yamlType: YamlType) => void;
```
Let's introduce a type for it, we can also use it on the state in `EditInstance`.

File: src/pages/instances/EditInstance.tsx (hunk 12 above)
Comment (edlerd) [reply src/pages/instances/EditInstance.tsx]: There is also a bug here currently. If we switch to `expandInheritedValues` and then leave the yaml view, the save button will be hidden.

File: src/util/yaml.tsx [hunk 14]
```diff
@@ -7,3 +9,132 @@ export const yamlToObject = (yamlString: string): object => {
...
+              const source = getSourceProfile(
+                reversedProfiles,
+                profiles,
+                propKey,
+                name,
+              );
+              const lines = propYaml.split("\n");
+              lines[0] = `${lines[0]} # inherited from profile: ${source}`;
+              propYaml = lines.join("\n");
+            }
```
Comment (edlerd) [inline src/util/yaml.tsx]: I doubt that is possible. Either the device is inherited, or it is locally overridden or locally defined. But single properties on a device should never come from another source as the device itself.

File: src/util/yaml.tsx [hunk 15]
```diff
@@ -7,3 +9,132 @@ export const yamlToObject = (yamlString: string): object => {
...
+    if (deviceName) {
+      const device = p?.devices?.[deviceName];
+      return (
+        device &&
+        (device as unknown as Record<string, unknown>)[key] !== undefined
+      );
+    }
+    return p?.config?.[key] !== undefined;
+  });
+  return source || "profile";
```
Comment (edlerd) [inline src/util/yaml.tsx]: I find this difficult to read and suspect a for loop over `allProfiles` with an early exit on match might be simpler. Also, I think we wouldn't need to reverse the profiles with that approach. Wdyt?

**Review by edlerd — COMMENTED** [review]
This is so useful to discover misconfigurations! Excellent work. Some edges are a bit too sharp, but I think with minor changes it should be good to merge.

File: src/pages/instances/EditInstance.tsx (hunk 11 above)
Comment (kimanhou) [reply src/pages/instances/EditInstance.tsx:297]: Do you want to align the YamlSwitch on the right ?
image

Comment (edlerd) [reply src/pages/instances/EditInstance.tsx:297]: Sorry, this was a mix-up on my side. I thought removing the boolean would simplify it, but it actually sets it to false, so it is needed. In my mind this was `isAlignLeft` and then the shorthand would work. 

Let's keep it as is.

File: src/components/forms/YamlConfirmation.tsx [hunk 16]
```diff
@@ -16,8 +21,8 @@ const YamlConfirmation: FC<Props> = ({ onConfirm, close }) => {
       title="Confirm"
     >
       <p>
-        Switching back to guided forms will discard all changes in the YAML
-        editor.
+        Switching {isExpanded ? "back to guided forms" : "to expanded view"}{" "}
+        will discard all changes in the YAML editor.
```
Comment (edlerd) [inline src/components/forms/YamlConfirmation.tsx]:
```suggestion
        This will discard all changes in the YAML editor.
```
Maybe simplify the message and keep the interface as it was. Wdyt?

File: src/components/forms/YamlTypeSelector.tsx [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+import { RadioInput, usePortal } from "@canonical/react-components";
+import YamlConfirmation from "components/forms/YamlConfirmation";
+import type { InstanceAndProfileFormikProps } from "types/forms/instanceAndProfileFormProps";
+
+export type YamlType = "local" | "expandInheritedValues";
+
+interface Props {
+  yamlType: YamlType;
+  setYamlType: (yamlType: YamlType) => void;
+  formik?: InstanceAndProfileFormikProps;
```
Comment (edlerd) [inline src/components/forms/YamlTypeSelector.tsx]:
```suggestion
  formik: InstanceAndProfileFormikProps;
```
Formik should be mandatory here, no?

File: src/components/forms/YamlTypeSelector.tsx [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+  const handleConfirmYamlTypeChange = () => {
+    if (formik && pendingYamlType) {
+      void formik.setFieldValue("yaml", undefined);
+      if (pendingYamlType === "expandInheritedValues") {
+        void formik.setFieldValue("readOnly", true);
+      }
+      setYamlType(pendingYamlType);
+    }
+    setPendingYamlType(null);
+    closePortal();
```
Comment (edlerd) [inline src/components/forms/YamlTypeSelector.tsx]:
```suggestion
    void formik.setFieldValue("yaml", undefined);
    setYamlType(pendingYamlType as YamlType);
    setPendingYamlType(null);
    closePortal();
```
A bit more straight forward?

Also the read only seems to be unused?

File: src/components/forms/YamlTypeSelector.tsx [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    if (formik) {
+      if (newYamlType === "expandInheritedValues") {
+        void formik.setFieldValue("yaml", undefined);
+        void formik.setFieldValue("readOnly", true);
+      }
+    }
+
+    setYamlType(newYamlType);
```
Comment (edlerd) [inline src/components/forms/YamlTypeSelector.tsx]:
```suggestion
    void formik.setFieldValue("yaml", undefined);
    setYamlType(newYamlType);
```
A bit simpler? This seems to have a large overlap with `handleConfirmYamlTypeChange` so maybe we can extract the general function and reuse it.

Also read only seems to be unused?

File: src/util/yaml.tsx [hunk 20]
```diff
@@ -7,3 +9,119 @@ export const yamlToObject = (yamlString: string): object => {
...
+  for (const pName of profileNames) {
+    const profile = allProfiles.find((profile) => profile.name === pName);
+    if (!profile) continue;
+
+    if (deviceName) {
+      const device = profile.devices?.[deviceName];
+      if (
+        device &&
+        (device as unknown as Record<string, unknown>)[key] !== undefined
+      ) {
```
Comment (edlerd) [inline src/util/yaml.tsx]:
```suggestion
      if (device) {
```
I think we don't care about the `type` key on the device here. If the device is found, we found the source of it.

File: src/util/yaml.tsx [hunk 21]
```diff
@@ -7,3 +9,119 @@ export const yamlToObject = (yamlString: string): object => {
...
+  if (Object.keys(expandedConfig).length > 0) {
+    sections.push("config:");
+    Object.entries(expandedConfig).forEach(([key, value]) => {
+      const isLocal = Object.prototype.hasOwnProperty.call(localConfig, key);
+
+      if (!isLocal && (value === "" || value === null)) {
+        const source = getSourceProfile(reversedProfiles, profiles, key);
+        sections.push(`  ${key}: "" # inherited from profile: ${source}`);
+        return;
+      }
```
Comment (edlerd) [inline src/util/yaml.tsx]:
```suggestion
```
Does this case exist? I think the value is either set or unset, but never empty.

File: src/components/forms/YamlTypeSelector.tsx (hunk 18 above)
Comment (kimanhou) [reply src/components/forms/YamlTypeSelector.tsx]: The read only is necessary for hiding the `Cancel` and `Save` buttons when switching to expanded mode.

Comment (edlerd) [reply src/components/forms/YamlTypeSelector.tsx]: Ah, I see, then let's keep it in :+1:

File: src/util/yaml.tsx [hunk 22]
```diff
@@ -7,3 +9,109 @@ export const yamlToObject = (yamlString: string): object => {
...
+    } else {
+      if (profile.config?.[key] !== undefined) {
+        return pName;
+      }
+    }
+  }
+  return "profile";
+};
+
+export const expandInheritedValuesYaml = (
```
Comment (edlerd) [inline src/util/yaml.tsx:37]: Can we add a unit sec test for this with the main cases?

I think that will help maintain this in the long term. Especially if we discover edge cases to add to the test suite, the newly discovered case, and making sure we don't break something accidentally.

File: src/components/forms/YamlTypeSelector.tsx [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+  yamlType: YamlType;
+  setYamlType: (yamlType: YamlType) => void;
+  formik: InstanceAndProfileFormikProps;
+}
+
+const YamlTypeSelector: FC<Props> = ({ yamlType, setYamlType, formik }) => {
+  const { openPortal, closePortal, isOpen, Portal } = usePortal();
+  const [pendingYamlType, setPendingYamlType] = useState<YamlType | null>(null);
+
+  const performYamlTypeSwitch = (newYamlType: YamlType) => {
```
Comment (edlerd) [inline src/components/forms/YamlTypeSelector.tsx]:
```suggestion
  const performSwitch = (newYamlType: YamlType) => {
```

File: src/components/forms/YamlTypeSelector.tsx [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+    void formik.setFieldValue("yaml", undefined);
+
+    if (newYamlType === "expandInheritedValues") {
+      void formik.setFieldValue("readOnly", true);
+    }
+
+    setYamlType(newYamlType);
+  };
+
+  const handleYamlTypeChange = (
```
Comment (edlerd) [inline src/components/forms/YamlTypeSelector.tsx]:
```suggestion
  const handleChange = (
```

File: src/components/forms/YamlTypeSelector.tsx [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+    if (formik.values.yaml) {
+      setPendingYamlType(newYamlType);
+      openPortal(e);
+      return;
+    }
+
+    performYamlTypeSwitch(newYamlType);
+  };
+
+  const handleConfirmYamlTypeChange = () => {
```
Comment (edlerd) [inline src/components/forms/YamlTypeSelector.tsx]:
```suggestion
  const handleConfirm = () => {
```

File: src/components/forms/YamlTypeSelector.tsx [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+    if (pendingYamlType) {
+      performYamlTypeSwitch(pendingYamlType);
+    }
+    setPendingYamlType(null);
+    closePortal();
+  };
+
+  const handleCancelYamlTypeChange = () => {
```
Comment (edlerd) [inline src/components/forms/YamlTypeSelector.tsx]:
```suggestion
  const handleCancel = () => {
```

File: src/util/yaml.tsx [hunk 27]
```diff
@@ -7,3 +9,109 @@ export const yamlToObject = (yamlString: string): object => {
...
   return dumpYaml(obj, { lineWidth: -1 });
 };
+
+const getSourceProfile = (
+  profileNames: string[],
+  allProfiles: LxdProfile[],
+  key: string,
+  deviceName?: string,
+): string => {
+  for (const pName of profileNames) {
```
Comment (edlerd) [inline src/util/yaml.tsx]:
```suggestion
  for (const profileName of profileNames) {
```

File: src/util/yaml.tsx (hunk 22 above)
Comment (kimanhou) [reply src/util/yaml.tsx:37]: I added some tests but the formatting can be a bit different

File: src/util/yaml.spec.ts [hunk 28]
```diff
@@ -0,0 +1,130 @@
...
+    );
+  });
+
+  it("should clean up excessive newlines while keeping padding for blocks", () => {
+    const result = expandInheritedValuesYaml(mockInstance, mockProfiles);
+    // Ensure we don't have triple newlines from the join/padding logic
+    expect(result).not.toMatch(/\n{3,}/);
+    // Ensure padding exists after the cloud-init block
+    expect(result).toMatch(/package_update: true\n\s*\n\s*/);
+  });
```
Comment (edlerd) [inline src/util/yaml.spec.ts]: The cases are all using the same data and not really covering the cases. Like the one below, does not truly ensure there is no inheritance comment on the line:
```
    // Local config should not have a comment
    expect(result).toContain("user.meta: local-value");
```
How about a slightly different approach with having the expected YAML output here in the test file? Then a single test case can match the output of the function against that.

File: src/util/yaml.spec.ts [hunk 29]
```diff
@@ -0,0 +1,130 @@
+import { expandInheritedValuesYaml } from "./yaml";
+import type { LxdInstance } from "types/instance";
+import type { LxdProfile } from "types/profile";
+
+describe("expandInheritedValuesYaml", () => {
+  const mockProfiles: LxdProfile[] = [
```
Comment (edlerd) [inline src/util/yaml.spec.ts]:
```suggestion
  const profiles: LxdProfile[] = [
```

File: src/util/yaml.spec.ts [hunk 30]
```diff
@@ -0,0 +1,130 @@
...
+          nictype: "bridged",
+          parent: "lxdbr0",
+          type: "nic",
+          network: "lxdbr0",
+        },
+      },
+    },
+  ];
+
+  const mockInstance: LxdInstance = {
```
Comment (edlerd) [inline src/util/yaml.spec.ts]:
```suggestion
  const instance: LxdInstance = {
```

**Review by edlerd — DISMISSED** [review]
LGTM, thanks for adding the tests. One idea for the test, though good to go as is from my side :)
